### DISCUSSION TITLE
API Plans?

### DISCUSSION THREAD

**Discussion by qdot** [runtime-shady-backroom/buttplug-lite#11] [Q&A] [answered]
Hey! buttplug.io author here. 

I'm currently compiling a list of repos using the library (https://github.com/buttplugio/awesome-buttplug) and happen to run into this. Funny enough, I actually need something exactly like this for NeosVR support, as it's REALLY difficult to form JSON in their in-world language, so having a simpler string parsing system would be handy.

Just curious what your plans are for using the simplified interface? Would be happy to discuss here, or if you want to jump on our discord, it's at https://discord.buttplug.io

Also: Thanks for using the library! If you have any input please let me know, almost everyone uses the FFI so I'm always looking for feedback on the Rust ergonomics.

**Comment by runtime-shady-backroom** [comment, marked as answer]
Huh, small world. The specific use case that inspired me to create this was Neos VR. I've seen an implementation of a straight buttplug.io client in Neos, and while "parsing" JSON using only String.IndexOf and String.Substring is a spectacle, it is not something I wanted any part in. So I built this project instead.

I've used Logix to hook a VirtualHapticPointSampler component to my intiface-proxy websocket, which makes my device behave exactly like a haptic device with real Neos support (e.g. haptic suits). You can make custom objects that drive the haptics system, with all the entertaining interactions that entails. I'd love to have the Neos side of the implementation in git as well, but unfortunately the only real way of sharing logix outside of Neos is via screenshots and videos.

My goals for this project are:
- The client shouldn't have to parse anything. Simple fire-and-forget commands that the server acts on if applicable are sufficient for my simple use case.
- The commands should be short and simple to avoid serialization complexity and reduce unneeded network usage. However, the commands also need to be strings as it's the only websocket data type Neos supports presently.

I'm still rather new to Rust, and therefore I'm not yet in a position to form meaningful opinions on Rust ergonomics. Rust specifics aside, the buttplug.io API seems well built and didn't give me any particular trouble. The only lingering doubt I have is when, if ever, I should stop scanning for devices. That the API lets me start and stop device scans at all seems to imply that there may be some hidden cost to scanning that perhaps I should be worried about.

Also, thanks for creating buttplug.io! I fully expected to have an absolutely awful time trying to interface with niche proprietary hardware, and finding a well-built open source project was a very welcome surprise.
